src/trie2.rs

```diff
@@ -0,0 +1,469 @@
+//! Binary Hash Tries, for representing sets and finite maps.
```

Comment:
r @kyleheadley 
This is the code that I really want reviewed.  The other `trie` variant (`trie.rs` and `trie1.rs`) above could probably be deleted.